File: scripts/update_benchmarks.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const SYMBOLS = ['SPY', 'QQQ', 'SPXL', 'TQQQ'];
const OUTPUT_DIR = path.join(__dirname, '../public/benchmarks');

/**
 * Fetches historical daily OHLC data from Yahoo Finance.
 * URL format: https://query1.finance.yahoo.com/v7/finance/download/{symbol}?period1={start}&period2={end}&interval=1d&events=history
 */
async function updateBenchmark(symbol) {
  const now = Math.floor(Date.now() / 1000);
  const start = Math.floor(new Date('2024-01-01').getTime() / 1000);
  
  const url = `https://query1.finance.yahoo.com/v7/finance/download/${symbol}?period1=${start}&period2=${now}&interval=1d&events=history`;
  
  console.log(`[${symbol}] Fetching data...`);
  
  try {
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }
    
    const csvContent = await response.text();
    const filePath = path.join(OUTPUT_DIR, `${symbol}.csv`);
    
    if (!fs.existsSync(OUTPUT_DIR)) {
      fs.mkdirSync(OUTPUT_DIR, { recursive: true });
    }
    
    fs.writeFileSync(filePath, csvContent);
    console.log(`[${symbol}] Saved to ${filePath}`);
  } catch (error) {
    console.error(`[${symbol}] Error:`, error.message);
  }
}

async function run() {
  console.log('--- Benchmarks Update Started ---');
  for (const symbol of SYMBOLS) {
    await updateBenchmark(symbol);
    // Add small delay to avoid rate limiting
    await new Promise(resolve => setTimeout(resolve, 1000));
  }
  console.log('--- Benchmarks Update Finished ---');
}

run();

File: scripts/update_benchmarks.py
import yfinance as yf
import pandas as pd
import os
from datetime import datetime

SYMBOLS = ['SPY', 'QQQ', 'SPXL', 'TQQQ', 'GRNY']
# Use absolute path to ensure it works from any directory (especially in GitHub Actions)
BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
OUTPUT_DIR = os.path.join(BASE_DIR, 'public/benchmarks')

def update_benchmarks():
    print(f"--- Benchmarks Update Started ({datetime.now()}) ---")
    
    if not os.path.exists(OUTPUT_DIR):
        os.makedirs(OUTPUT_DIR)

    for symbol in SYMBOLS:
        print(f"[{symbol}] Fetching data...")
        try:
            # Fetch from 2024-01-01 to now
            data = yf.download(symbol, start="2024-01-01", progress=False)
            
            if data.empty:
                print(f"[{symbol}] No data found.")
                continue

            # yfinance v0.2.x+ returns a MultiIndex for columns if one symbol is passed in some contexts.
            # We flatten it to just the OHLC columns.
            if isinstance(data.columns, pd.MultiIndex):
                data.columns = data.columns.get_level_values(0)

            # Keep only the columns we need
            df = data[['Open', 'High', 'Low', 'Close']].copy()
            
            # Save to CSV - ensuring a clean simple header: Date,Open,High,Low,Close
            file_path = os.path.join(OUTPUT_DIR, f"{symbol}.csv")
            df.to_csv(file_path)
            print(f"[{symbol}] Saved {len(df)} rows to {file_path}")
            
        except Exception as e:
            print(f"[{symbol}] Error: {str(e)}")

    print("--- Benchmarks Update Finished ---")

if __name__ == "__main__":
    update_benchmarks()
